### PR TITLE
sandbox: introduce the MKOSI_CHROOT_SUPPRESS_CHOWN variable

### PR DESCRIPTION
**Pull request by behrmann**
This allows suppressing chowns in scripts.

### CODE REVIEW

File: mkosi/resources/mkosi.md [hunk 1]
```diff
@@ -2730,6 +2730,10 @@ Note that the minimum required Python version is 3.9.
   changing ownership to root will work but changing ownership to any other user
   or group will fail.
 
+  Note that chown calls are only suppressed when running package managers, but
+  not during scripts. If this should be required, e.g. for a build script, you
```
Comment (DaanDeMeyer) [inline mkosi/resources/mkosi.md:2734]: "during scripts" => "when running scripts", "should be required"' => "is required"

File: mkosi/resources/mkosi.md [hunk 2]
```diff
@@ -2730,6 +2730,10 @@ Note that the minimum required Python version is 3.9.
...
+  can set the `MKOSI_CHROOT_SUPPRESS_CHOWN` variable to a non-empty value.
```
Comment (DaanDeMeyer) [inline mkosi/resources/mkosi.md:2735]: Please check for `1` or a boolean value instead of non-empty

File: mkosi/sandbox/__init__.py [hunk 3]
```diff
@@ -619,7 +619,8 @@ def main() -> None:
     upperdir = ""
     workdir = ""
     chdir = None
-    become_root = suppress_chown = unshare_net = unshare_ipc = False
+    become_root = unshare_net = unshare_ipc = False
+    suppress_chown = bool(os.getenv("MKOSI_CHROOT_SUPPRESS_CHOWN", False))
```
Comment (DaanDeMeyer) [inline mkosi/sandbox/__init__.py:623]: This should be processed outside of mkosi-sandbox. mkosi should look for the environment variable and pass --suppress-chown to chroot_cmd() when running scripts if it is set.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: mkosi/__init__.py [hunk 4]
```diff
@@ -568,8 +571,23 @@ def script_maybe_chroot_sandbox(
...
     helpers = {
-        "mkosi-chroot": chroot_script_cmd(tools=bool(context.config.tools_tree), network=network, work=True),
+        "mkosi-chroot": [
+            choose_interpreter(bool(context.config.tools_tree)), "-SI", "/sandbox.py",
+            "--bind", "/buildroot", "/",
+            "--bind", "/var/tmp", "/var/tmp",
+            *apivfs_options(root=Path("/")),
+            *chroot_options(),
+            "--bind", "/work", "/work",
+            "--chdir", "/work/src",
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:586]: You can pass `options` here as well and get rid of passing `--chdir` and `--suppress-chown` manually here

File: mkosi/run.py [hunk 5]
```diff
@@ -616,15 +603,12 @@ def chroot_cmd(
...
-        "python3" if tools or not exe.is_relative_to("/usr") else exe, "-SI", "/sandbox.py",
-        "--bind", "/buildroot", "/",
-        "--bind", "/var/tmp", "/var/tmp",
-        *apivfs_options(root=Path("/")),
-        *chroot_options(),
-        *(["--bind", "/work", "/work", "--chdir", "/work/src"] if work else []),
-        *(["--ro-bind-try", "/etc/resolv.conf", "/etc/resolv.conf"] if network else []),
-        "--",
-    ]
+def choose_interpreter(tools: bool) -> str:
```
Comment (DaanDeMeyer) [inline mkosi/run.py:606]: `finalize_interpreter`? Seems more consistent with how we usually name things

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Looks good, two minor things

File: mkosi/__init__.py (hunk 4 above)
Comment (behrmann) [reply mkosi/__init__.py]: Split out options from the caller and the ones here, which then allows the reuse.

File: mkosi/__init__.py [hunk 6]
```diff
@@ -566,10 +569,23 @@ def script_maybe_chroot_sandbox(
     options: Sequence[PathString],
     network: bool,
 ) -> Iterator[list[PathString]]:
-    options = ["--dir", "/work/src", "--chdir", "/work/src", *options]
+    localoptions = ["--dir", "/work/src", "--chdir", "/work/src"]
+
+    if parse_boolean(context.config.environment.get("MKOSI_CHROOT_SUPPRESS_CHOWN", "0")):
+        localoptions += ["--suppress-chown"]
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:575]: So this approach actually means we pass --suppress-chown to the non-chroot script as well. So we should handle this separately after all.

So I'd drop `localoptions` again and handle it like you did earlier after all, my bad.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: mkosi/__init__.py [hunk 7]
```diff
@@ -568,8 +571,23 @@ def script_maybe_chroot_sandbox(
 ) -> Iterator[list[PathString]]:
     options = ["--dir", "/work/src", "--chdir", "/work/src", *options]
 
+    suppress_chown = parse_boolean(context.config.environment.get("MKOSI_CHROOT_SUPPRESS_CHOWN", "0"))
+    if suppress_chown:
+        options += ["--suppress-chown"]
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:576]: This still passes `--suppress-chown` to the non-chroot command. This should only be added to the `mkosi-chroot` command options and the chroot script options.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: mkosi/__init__.py [hunk 8]
```diff
@@ -567,15 +570,30 @@ def script_maybe_chroot_sandbox(
...
+            "--",
+        ],
         "mkosi-as-caller": mkosi_as_caller(),
         **context.config.distribution.package_manager(context.config).scripts(context),
     }
 
     with finalize_host_scripts(context, helpers) as hd:
         if script.suffix != ".chroot":
+            if suppress_chown:
+                options += ["--suppress-chown"]
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:595]: This needs to be in the other branch

**Review by DaanDeMeyer — APPROVED** [review]

**Pull request merged**